File: src/main/scala/com/abaddon83/cardsGames/burracoGames/domainModels/burracoGames/initialised/mazzettos/MazzettoDeck.scala
package com.abaddon83.cardsGames.burracoGames.domainModels.burracoGames.initialised.mazzettos

import com.abaddon83.cardsGames.shares.decks.{Card, Deck}

import scala.collection.mutable.ListBuffer

case class MazzettoDeck private(override protected val cards: ListBuffer[Card]) extends Deck{

  override def grabFirstCard(): Card = {
    throw new UnsupportedOperationException("This method is not implemented in the Mazzetto")
  }

  override def grabAllCards(): List[Card] = {
    throw new UnsupportedOperationException("This method is not implemented in the Mazzetto")
  }

  def getCards(): List[Card] = {
    cards.toList
  }


}


object MazzettoDeck{
  def build(cards:List[Card]): MazzettoDeck = {
    assert(cards.size == 11 || cards.size == 18, "Pozzetto Size wrong")
    new MazzettoDeck(ListBuffer.from(cards))
  }
}

File: src/main/kotlin/com/abaddon83/burracoGame/domainModels/MazzettoDeck.kt
package com.abaddon83.burracoGame.domainModels

import com.abaddon83.burracoGame.shared.decks.Card
import com.abaddon83.burracoGame.shared.decks.Deck

data class MazzettoDeck private constructor(override val cards: MutableList<Card>) : Deck {

    override fun grabFirstCard(): Card = throw UnsupportedOperationException("This method is not implemented in the Mazzetto")

    override fun grabAllCards(): List<Card> = throw UnsupportedOperationException("This method is not implemented in the Mazzetto")

    fun getCardList(): List<Card> = cards.toList()

    companion object Factory {
        fun create(cards:List<Card>): MazzettoDeck {
            require(cards.size == 11 || cards.size == 18){"Pozzetto Size wrong, current size: ${cards.size}"}
            return MazzettoDeck(cards.toMutableList())
        }

    }

}